SSA/Projects/LLVMRiscV/Pipeline/Combiners.lean

```diff
@@ -0,0 +1,446 @@
+import SSA.Projects.LLVMRiscV.PeepholeRefine
+import SSA.Projects.LLVMRiscV.simpproc
+import SSA.Projects.RISCV64.Tactic.SimpRiscV
+import SSA.Projects.LLVMRiscV.Pipeline.mkRewrite
+
+open LLVMRiscV
+
+/- This file implements `DAGCombiner` patterns extrcted from the LLVM Risc-V backend.
+  First, we implement the Lean structure that implements the rewrite patterns and then we implement
+  optimizations for LLVM IR and RISC-V.
+  In particular, we implement the patterns supported by LLVM's `GlobalIsel` for RISC-V.
+  Because `GlobalIsel` is hybrid, some of these patterns regard generic IR,
+  while some are target-dependent.
+-/
+
+@[simp_riscv] lemma toType_bv : TyDenote.toType (Ty.riscv (.bv)) = BitVec 64 := rfl
+@[simp_riscv] lemma id_eq1 {α : Type} (x y : α) :  @Eq (Id α) x y = (x = y):= by simp only
+
+structure RISCVPeepholeRewrite (Γ : Ctxt Ty) where
+  lhs : Com LLVMPlusRiscV Γ .pure (Ty.riscv (.bv))
+  rhs : Com LLVMPlusRiscV Γ .pure (Ty.riscv (.bv))
+  correct :  (lhs.denote) = (rhs.denote) := by
+    simp_lowering
+    bv_decide
+
+/-! # Post-legalization optimizations
+
+  We implement post-legalization optimizations from LLVM's `GlobalISel` instructor selector.
+  Our naming conventions are consistent with the RISC-V backend.
+
+  We do not support known-bits analysis nor matching on values, and therefore do not implement the
+  patterns relying on this infrastructure (e.g. `shift_immed_chain`).
+-/
+
+/-- ### sub_to_add
+  (sub x, C) → (add x, -C)
+-/
+def sub_to_add_pat : LLVMPeepholeRewriteRefine 64 [Ty.llvm (.bitvec 64)] where
+  lhs := [LV| {
+  ^entry (%x: i64):
+  %c = llvm.mlir.constant 1 : i64
+  %1 = llvm.sub %x, %c : i64
+  llvm.return %1 : i64
```

Comment:
```suggestion
  ^entry (%x: i64):
    %c = llvm.mlir.constant 1 : i64
    %1 = llvm.sub %x, %c : i64
    llvm.return %1 : i64
```

This is probably super nitpicky (and I am happy to leave it as is), but LLVM would indent after the basic block.